### PR TITLE
feat: support Spanish emojis

### PR DESCRIPTION
**Pull request by dummy-1111**
### Fixed Issues

$ https://github.com/Expensify/App/issues/16086
PROPOSAL: https://github.com/Expensify/App/issues/16086#issuecomment-1545263685

### Tests
1. Open NewDot and go to any chat
2. Hovering any message shows reaction bar. Hover any emoji on the reaction bar and verify that emoji shortcode is translated
3. Hover any emoji reactions under the message and verify that the tooltip shows translated emoji short code. On mobile long press any emoji reactions and verify
4. Clicking the emoji+ button shows the emoji picker popup.
5. In this popup, verify that
    - search by Spanish keyword works
    - hovering category icons shows translated category names
    - skintone selection works for emojis supporting skintone
    - selecting emoji toggles emoji reaction
6. In the composer, verify that emoji suggestion and replacing works
7. Clicking the emoji button in the composer shows the emoji picker popup. Verify all items in step 5

- [x] Verify that no errors appear in the JS console

### Offline tests
1. Open NewDot and go to any chat
2. Hovering any message shows reaction bar. Hover any emoji on the reaction bar and verify that emoji shortcode is translated
3. Hover any emoji reactions under the message and verify that the tooltip shows translated emoji short code. On mobile long press any emoji reactions and verify
4. Clicking the emoji+ button shows the emoji picker popup.
5. In this popup, verify that
    - search by Spanish keyword works
    - hovering category icons shows translated category names
    - skintone selection works for emojis supporting skintone
    - selecting emoji toggles emoji reaction
6. In the composer, verify that emoji suggestion and replacing works
7. Clicking the emoji button in the composer shows the emoji picker popup. Verify all items in step 5

### QA Steps
1. Open NewDot and go to any chat
2. Hovering any message shows reaction bar. Hover any emoji on the reaction bar and verify that tooltip is translated in preferred language
3. Hover any emoji reactions under the message and verify that the tooltip shows translated emoji short code. On mobile long press any emoji reactions and verify
4. Clicking the emoji+ button shows the emoji picker popup.
5. In this popup, verify that
    - search by Spanish keyword works when preferred locale is English
    - search by Spanish and English keyword works when preferred locale is Spanish
    - hovering category icons shows translated category names
    - skintone selection works for emojis supporting skintone
    - selecting emoji toggles emoji reaction
6. In the composer, verify that emoji suggestion and replacing works
7. Clicking the emoji button in the composer shows the emoji picker popup. Verify all items in step 5
8. Run all reaction and emoji picker related regressions

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/126839717/e449f7c6-40b8-42cf-a48b-e9009fa49618

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/126839717/6a41a03a-f13f-476b-bdc2-d68ddcdb9a33

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/126839717/f74646c1-b9fc-4d5a-a6f4-6fd0e2d234a2

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/126839717/1020d562-2ab6-4ab3-b2da-b1916e899b69

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/126839717/3128804c-02a3-4240-ad8a-1acf1549c391

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/126839717/556d9115-6267-435c-bc8a-903320c1b912

</details>

### CODE REVIEW

File: src/CONST.js
Comment (s-alves10) [reply src/CONST.js:1334]: This is for clarity. They are not names of emojis, just short codes of them. I mentioned this in issue before

File: src/components/EmojiPicker/CategoryShortcutBar.js
Comment (s-alves10) [reply src/components/EmojiPicker/CategoryShortcutBar.js:15]: Headers has no code unlike emojis. It has only name and icon. I wanted to add meanings to the properties

**Review by 0xmiroslav — COMMENTED** [review]
Your updated structure of name -> shortcode, header code -> name can be follow-up if needed.
In this PR, let's reduce code diff as much as possible and try to keep original structure but just add localization.

File: src/components/EmojiPicker/CategoryShortcutBar.js
Comment (s-alves10) [reply src/components/EmojiPicker/CategoryShortcutBar.js:15]: I'll change the headers back to original Spanish.
But let's keep the current emoji data structure as is. We have all Spanish emojis in its own structure, and we treat headers as emojis as well. Why don't we translate headers in the same place? I don't understand why headers should be in en/es.js. We can safely remove all emojis from en/es.js. This is more clear and reasonable, I think.

Comment (s-alves10) [reply src/components/EmojiPicker/CategoryShortcutBar.js:15]: I think we should think about extension of supported languages, not about original structure.

Comment (stitesExpensify) [reply src/components/EmojiPicker/CategoryShortcutBar.js:15]: I'm confused @s-alves10 what is the benefit of this change? In the future we're going to add other files like `de.js` which with have german translations etc. for the whole app, so it seems like keeping the current structure makes more sense than changing it

**Review by stitesExpensify — CHANGES_REQUESTED** [review]
I'm a bit confused about some of these changes. I was under the impression that we were just adding spanish keywords to the file, and creating a spanish Trie. If you could clarify why other changes were necessary that would be great @s-alves10

**Comment by s-alves10** [comment]
@stitesExpensify @0xmiroslav 

Thank you for your feedback
Let me explain.

Emojis have its own data structure supporting multiple languages.
1. In order to storing all translations regarding emojis with the same structure, I changed quick emojis as well. Those have the same structure as normal emojis.
2. The order of skintones is different from what we see in the ND. It makes me confused and would make other developers as well. That's why I changed its order in the code
3. Supporting more languages isn't a problem in our current structure. All emoji translations would be removed under languages folder and we can care only other translations. The resource I used supports many other languages. Separating emojis and others in translation isn't more convenient?
4. Changing emoji data structure leads to a little more changes in the project but those are required in order to keep our emojis data consistently over the entire project.

**Comment by stitesExpensify** [comment]
I agree. I don't think that we need to make such big changes in this PR. All we really want from this PR is for typing `:sonrisa:` to come up with 😄 and everything else to stay the same

File: src/CONST.js
Comment (s-alves10) [reply src/CONST.js:1336]: As you can see below, skintone numbering is weird. I think this change is reasonable.

https://github.com/s-alves10/App/blob/83ab0160a3b5c385fd26fffe20bdbccabf9249c0/assets/emojis.js#L56-L81

**Comment by s-alves10** [comment]
@0xmiroslav @stitesExpensify 

I changed the emoji data structure back to original. Now name property is object instead of string. Only necessary changes were left. Please check

**Comment by s-alves10** [comment]
> @s-alves10 just to confirm: you implemented [this suggestion](https://github.com/Expensify/App/issues/16086#issuecomment-1595369534) or not?

I was waiting for your suggestion. Do you want this feature to be implemented?

**Review by 0xmiroslav — COMMENTED** [review]
This is edge case but I am not sure if this can be bug or not (I don't think bug) but at least worth to consider:

(when switch language in another device)

https://github.com/Expensify/App/assets/97473779/2e7b0030-2dc7-4c99-b19d-b7755f639215

File: src/CONST.js
Comment (s-alves10) [reply src/CONST.js:1336]: Yeah. I'm sure

**Comment by s-alves10** [comment]
> This is edge case but I am not sure if this can be bug or not (I don't think bug) but at least worth to consider:
> 
> (when switch language in another device)
> 
>  Screen.Recording.2023-06-18.at.7.28.25.PM.mov

I think this is the expected behavior.

**Comment by s-alves10** [comment]
> Some emoji orders don't match with current version:

@0xmiroslav 

We're now using different database and the order is so important?

There are 1812 emojis in the old database and now 1811 emojis(except headers).
`🧑‍🎄` is missing in our current database because its Spanish version is missing.

**Comment by s-alves10** [comment]
> > 🧑‍🎄 is missing in our current database because its Spanish version is missing.
> 
> I don't want anyone to report this as regression. Let's imagine react with this emoji in old app and not see in new app. Actually crashes. If missing, why can't we add Spanish name and keywords manually? (at least fill Spanish values with English ones if hard to find Spanish copies)
> 
> Screenshot 2023-06-18 at 8 29 42 PM

Good idea. I'll do it

File: src/components/Reactions/ReportActionItemReactions.js
Comment (s-alves10) [reply src/components/Reactions/ReportActionItemReactions.js]: Done

File: src/libs/EmojiUtils.js
Comment (s-alves10) [reply src/libs/EmojiUtils.js]: Done

**Comment by s-alves10** [comment]
@0xmiroslav 

PR was updated
1. Missing emoji added with same name and keywords for both English and Spanish
2. `getEmojiName` function created in `EmojiUtils.js`
3. `getLocalizedEmojiName` optimzed
4. Use CONST.LOCALES.DEFAULT

**Comment by s-alves10** [comment]
That comment is for extracting the old emoji database, I think. Is that needed for new database?

**Comment by s-alves10** [comment]
I extracted the new database in several steps. Manual operations are even needed. We can specify like "This database is extracted from `URL`" but cannot add the whole extraction process. What can I do?

**Comment by s-alves10** [comment]
> Above database concerns should have been discussed earlier in proposal stage.

https://github.com/Expensify/App/issues/16086#issuecomment-1580123722
https://github.com/Expensify/App/issues/16086#issuecomment-1587803766

**Comment by s-alves10** [comment]
I think we don't need to concern about keywords so much as long as we support all emojis in the old database.
You can see `smile` keyword doesn't match grinning in Slack.

**Comment by s-alves10** [comment]
> Different keywords are fine but not about name. We should make it compatible with old version. It affects all users who already reacted emojis with different names before this PR.

What do you suggest in this case?

**Comment by s-alves10** [comment]
I have a suggestion
Let's create the emoji name like
```
{
  en: 'xxx',
  es: 'yyy',
  old: 'zzz',
}
```

And handle emojis with old names

**Comment by s-alves10** [comment]
We can use old names for English, new names for Spanish.

**Comment by s-alves10** [comment]
What do you mean Spanish old names?

**Comment by s-alves10** [comment]
Name is added to keywords automatically in Trie. You can test it

**Comment by stitesExpensify** [comment]
Okay, it looks like there has been a lot going on here and I think that we need to take a step back to set expectations and decide how to proceed. This conversation probably should have happened in the issue, but I think that this conversation is necessary right now.

Expectations:
1. You can search for emojis in the language you have chosen, as well as english
    - This should be extensible so that when we add more languages we don't need a lot of changes
2. The current structure of emoji.js with regards to `name` and `code` should not change
3. The english names of emojis should not be changed because we already have them saved in the DB
4. Skintone order should not be changed because users who have reacted with a certain skin tone will now have reacted with a different skin tone
5. All reactions should be sent to the server using the english name

Decisions:
1. Should we add more keywords to emoji.js, or create separate `emoijs/common.js` (for skintones), `en.emojis.js`, `es.emojis.js`?
    - Personally I am in favor of creating separate files so that we mirror `en.js` and `es.js`. This has the added benefit of not changing the existing emoji names which are already in the database.
2. How should we display the usages of the emojis? (Show it in chosen language always, or show in english + language, or whatever language was used
    - I think that chosen language makes the most sense


Are there any other decisions that need to be made here? cc @0xmiroslav

**Comment by s-alves10** [comment]
> We can use current names/keywords for English, new names/keywords for Spanish.

@0xmiroslav 

I think https://github.com/Expensify/App/pull/20828#issuecomment-1597677913 includes this.

**Comment by stitesExpensify** [comment]
> We can use current names/keywords for English, new names/keywords for Spanish.

Everything should be referenced / stored in the DB using the english name. 
For display on the front end we should be translating `english name` -> `code` -> `<language> name`

**Comment by s-alves10** [comment]
@0xmiroslav 

skintones, headers, types are the same for all locales. I think these should be stored in common.js

**Comment by s-alves10** [comment]
@0xmiroslav @stitesExpensify 

https://github.com/Expensify/App/pull/20828#issuecomment-1598235677 were all implemented
Please review

**Comment by stitesExpensify** [comment]
> just to confirm that this was your concern?

Yes, that is what I had in mind. The only potential issue is that when rendering we need the headers in the same list, so headers might actually have to be duplicated.

**Comment by stitesExpensify** [comment]
FYI I will be OOO until 6/27 so the reviews of this PR will be on hold for the next few days. If you think that this PR needs to be completed faster than that, feel free to reach out in #expensify-open-source for a new reviewer, or post on the issue and the BugZero team member can help out!

**Comment by s-alves10** [comment]
@0xmiroslav 

Please check the updated PR

**Comment by stitesExpensify** [comment]
Feel free to start testing when you get a chance @0xmiroslav, I'm going to do the same right now.

**Comment by stitesExpensify** [comment]
@iwiznia would you mind giving this a test to please since you actually use the product in Spanish? I just want to make sure that we are creating the expected UI, especially when doing `:colon-codes:`

**Comment by stitesExpensify** [comment]
@s-alves10 can you add to the QA steps that we should be testing this for:
1. Both english and spanish
2. Switching between the languages updates the tooltips
3. When set to spanish you can search in english, but not vice-versa

**Comment by s-alves10** [comment]
> @s-alves10 can you add to the QA steps that we should be testing this for:
> 
> 1. Both english and spanish
> 2. Switching between the languages updates the tooltips
> 3. When set to spanish you can search in english, but not vice-versa

I've updated the PR

**Comment by stitesExpensify** [comment]
cc @JmillsExpensify in case you want to give it a shot :)

**Review by 0xmiroslav — COMMENTED** [review]
Console error when type in emoji picker in android but not a blocker as already happening on production

console

**Review by 0xmiroslav — APPROVED** [review]
LGTM.
@stitesExpensify all yours!

**Review by 0xmiroslav — CHANGES_REQUESTED** [review]

File: src/components/Reactions/QuickEmojiReactions/BaseQuickEmojiReactions.js
Comment (s-alves10) [reply src/components/Reactions/QuickEmojiReactions/BaseQuickEmojiReactions.js:60]: Fixed

**Review by 0xmiroslav — APPROVED** [review]

**Review by 0xmiroslav — APPROVED** [review]

**Comment by s-alves10** [comment]
@stitesExpensify 

Will you take a look at the PR again?

**Comment by stitesExpensify** [comment]
Hi there! Taking a look!

**Review by stitesExpensify — CHANGES_REQUESTED** [review]
Code is looking great! Just a couple of small questions before I approve

File: src/libs/EmojiTrie.js
Comment (s-alves10) [reply src/libs/EmojiTrie.js:51]: In dev mode, these operations take about 500ms. In production would be far shorter.
These operations are based on the old function. Almost same.

**Comment by s-alves10** [comment]
@stitesExpensify 

Please take a look at the changes. Thank you

**Review by stitesExpensify — APPROVED** [review]

**Comment by jjcoffee** [comment]
Just coming from [this issue](https://github.com/Expensify/App/issues/25735) that fixed a bug that this PR introduced that caused duplicate suggestions to appear. Since we add the English Trie node with the suggestion's name in English, this won't match the equivalent Spanish emoji name, so it gets added as a separate match here:

https://github.com/Expensify/App/blob/3f9f1e6e2ce91468a536755f8d5830a2f77d7eaa/src/libs/EmojiUtils.js#L381

**Pull request merged**
